src/main/resources/db/changelog/20231003113535-entity_layering_schema.xml

```diff
@@ -354,7 +354,357 @@
         <dropForeignKeyConstraint baseTableName="cp_pool" constraintName="cp_pool_fk1"/>
     </changeSet>
 
-    <!-- TODO: Add migration logic here -->
+   <changeSet id="20231003113535-11" author="crog" dbms="postgresql, mysql, mariadb">
+        <sql>
+            -- Migrate global products and content
+            DROP TABLE IF EXISTS tmp_active_global_products;
+            CREATE TEMPORARY TABLE tmp_active_global_products (
+                uuid VARCHAR(32) NOT NULL,
+                pid VARCHAR(32) NOT NULL,
+                new_uuid VARCHAR(32) NOT NULL,
+                best BOOLEAN NOT NULL DEFAULT false,
+                PRIMARY KEY (uuid));
+
+            DROP TABLE IF EXISTS tmp_active_global_contents;
+            CREATE TEMPORARY TABLE tmp_active_global_contents (
+                uuid VARCHAR(32) NOT NULL,
+                cid VARCHAR(32) NOT NULL,
+                new_uuid VARCHAR(32) NOT NULL,
+                best BOOLEAN NOT NULL DEFAULT false,
+                PRIMARY KEY (uuid));
+
+            -- Collect "active" global products and their children
+            INSERT INTO tmp_active_global_products (uuid, pid, new_uuid, best)
+                WITH RECURSIVE parent_child_map(parent_uuid, child_uuid) AS (
+                    SELECT uuid AS parent_uuid, derived_product_uuid AS child_uuid
+                        FROM cp2_products
+                        WHERE derived_product_uuid IS NOT NULL
+                    UNION
+                    SELECT product_uuid AS parent_uuid, provided_product_uuid AS child_uuid
+                        FROM cp2_product_provided_products),
+                product_graph(uuid, depth) AS (
+                    SELECT active.uuid, 0
+                        FROM cp2_products prod
+                        JOIN (SELECT product_uuid AS uuid FROM cp2_owner_products
+                            UNION
+                            SELECT product_uuid AS uuid FROM cp_pool) active ON active.uuid = prod.uuid
+                        WHERE prod.locked != 0
+                    UNION
+                    SELECT pcmap.child_uuid, pg.depth + 1
+                        FROM parent_child_map pcmap
+                        JOIN product_graph pg ON pg.uuid = pcmap.parent_uuid
+                        -- shouldn't be needed, but here just in case someone decided to be cheeky and define a
+                        -- cyclical product, or some other crazy thing that's entirely unsupported by CP's logic
+                        -- and queries
+                        WHERE pg.depth &lt; 9)
+                SELECT prod.uuid, prod.product_id, md5(concat('--global--', prod.product_id)) AS new_uuid, prod.uuid = best.uuid AS best_uuid
+                    FROM (SELECT DISTINCT uuid FROM product_graph) active
+                    JOIN cp2_products prod ON prod.uuid = active.uuid
+                    JOIN (SELECT prod2.product_id, MAX(prod2.uuid) AS uuid
```

Comment:
is this MAX depending on the fact that our uuids are incremental (so a newer version always has a higher uuid)?

Reply:
Very good observation! This was why I picked it as the tie breaker, but that behavior isn't required for it. The inner query is our primary selector, using the most recent updated date for a given product ID (and owner ID + product ID in the custom block), only falling back to the `MAX(prod2.uuid)` bit in the unlikely case we have multiple entities with the exact same last update date.

In terms of how the query behaves, it's not required to pick the "newest" UUID, just that we pick a consistent UUID.

Reply:
As an added bit of history here: I think the tie breaker selection mattered more when I was manually doing that "best" lookup more than once. When I moved the flag to the active_custom and active_global tables, the consistency tie breaker probably became less valuable, but it doesn't hurt to retain it.